flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -236,6 +236,89 @@ SqlDrop SqlDropTable(Span s, boolean replace) :
     }
 }
 
+/**
+* Parses an INSERT statement.
+*/
+SqlNode RichSqlInsert() :
+{
+    final List<SqlLiteral> keywords = new ArrayList<SqlLiteral>();
+    final SqlNodeList keywordList;
+    SqlNode table;
+    SqlNodeList extendList = null;
+    SqlNode source;
+    final SqlNodeList partitionList = new SqlNodeList(getPos());
+    SqlNodeList columnList = null;
+    final Span s;
+}
+{
+    (
+        <INSERT>
+    |
+        <UPSERT> { keywords.add(SqlInsertKeyword.UPSERT.symbol(getPos())); }
+    )
+    { s = span(); }
+    SqlInsertKeywords(keywords) {
+        keywordList = new SqlNodeList(keywords, s.addAll(keywords).pos());
+    }
+    <INTO> table = CompoundIdentifier()
+    [
+        LOOKAHEAD(5)
+        [ <EXTEND> ]
+        extendList = ExtendList() {
+            table = extend(table, extendList);
+        }
+    ]
+    [
+        LOOKAHEAD(2)
+        { final Pair<SqlNodeList, SqlNodeList> p; }
+        p = ParenthesizedCompoundIdentifierList() {
+            if (p.right.size() > 0) {
+                table = extend(table, p.right);
+            }
+            if (p.left.size() > 0) {
+                columnList = p.left;
+            }
+        }
+    ]
+    [
+        <PARTITION> PartitionSpecCommaList(partitionList) {
+            if (this.conformance != FlinkSqlConformance.HIVE) {
```

Comment:
I think we should add a method to `SqlConformance` such as `isInsertPartitionAllowed`. And use the method here instead of hard code a specific HIVE dialect. In this way, we can extend other dialect which also supports insert partition.

Reply:
We can have a method like this, what is the best impl i want, but merging to Apache Calcite needs time.

Reply:
Oh, I mean add this method to `FlinkSqlConformance`.

Reply:
I think we can do this if we always use FlinkSqlConformance for our flink parser.